evidence/pool.go

```diff
@@ -0,0 +1,80 @@
+package evidence
+
+import (
+	"github.com/tendermint/tmlibs/log"
+
+	cfg "github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/types"
+)
+
+// EvidencePool maintains a pool of valid evidence
+// in an EvidenceStore.
+type EvidencePool struct {
+	config *cfg.EvidenceConfig
+	logger log.Logger
+
+	state         types.State
+	evidenceStore *EvidenceStore
+
+	evidenceChan chan types.Evidence
+}
+
+func NewEvidencePool(config *cfg.EvidenceConfig, evidenceStore *EvidenceStore, state types.State) *EvidencePool {
+	evpool := &EvidencePool{
+		config:        config,
+		logger:        log.NewNopLogger(),
+		evidenceStore: evidenceStore,
+		state:         state,
+		evidenceChan:  make(chan types.Evidence),
+	}
+	return evpool
+}
+
+// SetLogger sets the Logger.
+func (evpool *EvidencePool) SetLogger(l log.Logger) {
+	evpool.logger = l
+}
+
+// EvidenceChan returns an unbuffered channel on which new evidence can be received.
+func (evpool *EvidencePool) EvidenceChan() <-chan types.Evidence {
+	return evpool.evidenceChan
+}
+
+// PriorityEvidence returns the priority evidence.
+func (evpool *EvidencePool) PriorityEvidence() []types.Evidence {
+	return evpool.evidenceStore.PriorityEvidence()
+}
+
+// PendingEvidence returns all uncommitted evidence.
+func (evpool *EvidencePool) PendingEvidence() []types.Evidence {
+	return evpool.evidenceStore.PendingEvidence()
+}
+
+// AddEvidence checks the evidence is valid and adds it to the pool.
+// Blocks on the EvidenceChan.
+func (evpool *EvidencePool) AddEvidence(evidence types.Evidence) (err error) {
+
+	// XXX: is this thread safe ?
```

Comment:
State is not thread safe... it's actually a bit of a mess.
(e.g. we should probably move state.LoadValidators out into a global function and keep state a singleton snapshot).  Is AddEvidence thread safe?

Reply:
Ok. This is actually broken because this state never gets updated. The goal with evidence pool was to be thread safe thanks to the threadsafety of the underlying db. But with needing to update the state, we probably need something more